validator/sawtooth_validator/state/client_handlers.py

```diff
@@ -382,6 +385,103 @@ def id_by_index(index, resources):
             return resources[index].address
 
 
+class _Sorter(object):
+    """A static class containing a method to sort lists of resources based on
+    SortControls sent with the request.
+    """
+    @classmethod
+    def sort_resources(cls, request, resources, fail_enum, header_proto=None):
+        """Sorts a list of resources based on a list of sort controls
+
+        Args:
+            request (object): The parsed protobuf request object
+            resources (list of objects): The resources to be sorted
+            fail_enum (int, enum): The enum status to raise with invalid keys
+            header_proto(class): Class to decode a resources header
+
+        Returns:
+            list: The sorted list of resources
+        """
+        if not request.sorting:
+            return resources
+
+        value_handlers = [cls._ValueHandler(controls, fail_enum, header_proto)
+                          for controls in request.sorting]
+
+        def sorter(resource_a, resource_b, value_handlers=value_handlers):
+            if not value_handlers:
+                return 0
+            handler = value_handlers[0]
+
+            value_a, value_b = handler.get_sort_values(resource_a, resource_b)
+
+            if value_a < value_b:
+                return handler.xform_result(-1)
+            if value_a > value_b:
+                return handler.xform_result(1)
+            return sorter(resource_a, resource_b, value_handlers[1:])
```

Comment:
Beware, python does not optimize tail recursion: http://neopythonic.blogspot.com.au/2009/04/tail-recursion-elimination.html

So this is limited to lists of length ~1000. If I understand this right though, that is probably fine since it is unlikely there will be >1000 sort keys.

It could be a security issue though... A malicious actor could query a bunch of validators with requests that have a bunch of sort keys and bring them down. Is there a hard limit somewhere before this on the number of sort keys?